server/src/main/java/org/elasticsearch/search/sort/RawBucketedSort.java

```diff
@@ -0,0 +1,340 @@
+    private final SortOrder order;
+    private final int bucketSize;
+    /**
+     * {@code true} if the bucket is in heap mode, {@code false} if
+     * it is still gathering.
+     */
+    private final BitArray heapMode;
+
+    protected RawBucketedSort(BigArrays bigArrays, SortOrder order, int bucketSize) {
+        this.bigArrays = bigArrays;
+        this.order = order;
+        this.bucketSize = bucketSize;
+        heapMode = new BitArray(1, bigArrays);
+    }
+
+    /**
+     * The order of the sort.
+     */
+    public final SortOrder getOrder() {
+        return order;
+    }
+
+    /**
+     * The number of values to store per bucket.
+     */
+    public int getBucketSize() {
+        return bucketSize;
+    }
+
+    /**
+     * Get the bound indexes (inclusive, exclusive) of the values for a bucket.
+     * Returns [0, 0] if the bucket has never been collected.
+     * <p>
+     *     To be used by {@link #getValues(long)} implementations.
+     * </p>
+     */
+    protected final Tuple<Long, Long> getBucketValuesBounds(long bucket) {
+        long rootIndex = bucket * bucketSize + 1;
+        if (rootIndex >= values().size()) {
+            // We've never seen this bucket.
+            return Tuple.tuple(0L, 0L);
+        }
+        long start = inHeapMode(bucket) ? rootIndex : (rootIndex + getNextGatherOffset(rootIndex) + 1);
+        long end = rootIndex + bucketSize;
+        return Tuple.tuple(start, end);
+    }
+
+    /**
+     * Get the values for a bucket if it has been collected.
+     * If it hasn't, then returns an empty list.
+     */
+    public abstract List<T> getValues(long bucket);
+
+    /**
+     * Is this bucket a min heap {@code true} or in gathering mode {@code false}?
+     */
+    public boolean inHeapMode(long bucket) {
+        return heapMode.get(bucket);
+    }
+
+    /**
+     * The {@linkplain BigArray} backing this sort.
+     */
+    protected abstract BigArray values();
+
+    /**
+     * Grow the {@linkplain BigArray} backing this sort to account for new buckets.
+     * This will only be called if the array is too small.
+     */
+    protected abstract void growValues(long minSize);
+
+    /**
+     * Get the next index that should be "gathered" for a bucket rooted
+     * at {@code rootIndex}.
+     */
+    protected abstract int getNextGatherOffset(long rootIndex);
+
+    /**
+     * Set the next index that should be "gathered" for a bucket rooted
+     * at {@code rootIndex}.
+     */
+    protected abstract void setNextGatherOffset(long rootIndex, int offset);
+
+    /**
+     * {@code true} if the entry at index {@code lhs} is "better" than
+     * the entry at {@code rhs}. "Better" in this means "lower" for
+     * {@link SortOrder#ASC} and "higher" for {@link SortOrder#DESC}.
+     */
+    protected abstract boolean betterThan(long lhs, long rhs);
+
+    /**
+     * Swap the data at two indices.
+     */
+    protected abstract void swap(long lhs, long rhs);
```

Comment:
I think this is actually going to be slow and we'd be better off building this class with stringtemplate. Java's going to call this method with `invokevirtual` and it's not going to be able to unroll loops and stuff. Testing this is hard - you need three subclasses and you need to initialize all of them. You need to hit this method call on all of them, just to be sure. That's why the jmh tests do that funny ritual where they pre-check all configurations. It pollutes the class space to force the jvm to handle all of the types.